src/backend/opencl/convolve.cpp

```diff
@@ -103,5 +112,184 @@ INSTANTIATE(ushort, float)
 INSTANTIATE(short, float)
 INSTANTIATE(uintl, float)
 INSTANTIATE(intl, float)
+#undef INSTANTIATE
+
+template <typename T, typename accT>
+Array<T> convolve2_unwrap(const Array<T> &signal, const Array<accT> &filter,
+                          const dim4 stride, const dim4 padding,
+                          const dim4 dilation) {
+    dim4 sDims = signal.dims();
+    dim4 fDims = filter.dims();
+
+    dim_t outputWidth =
+        1 +
+        (sDims[0] + 2 * padding[0] - (((fDims[0] - 1) * dilation[0]) + 1)) /
+            stride[0];
+    dim_t outputHeight =
+        1 +
+        (sDims[1] + 2 * padding[1] - (((fDims[1] - 1) * dilation[1]) + 1)) /
+            stride[1];
+    dim4 oDims = dim4(outputWidth, outputHeight, fDims[3], sDims[3]);
+
+    const bool retCols    = false;
+    Array<accT> unwrapped = unwrap_dilated(
+        cast<accT>(signal), fDims[0], fDims[1], stride[0], stride[1],
+        padding[0], padding[1], dilation[0], dilation[1], retCols);
+
+    unwrapped  = reorder(unwrapped, dim4(1, 2, 0, 3));
+    dim4 uDims = unwrapped.dims();
+    unwrapped.modDims(dim4(uDims[0] * uDims[1], uDims[2] * uDims[3]));
+
+    Array<accT> collapsedFilter = filter;
+
+    vector<af_seq> flip_index(4);
+    af_seq s      = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s             = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
```

Comment:
Perhaps, this `genFlipIndex` helper I suggested earlier should be moved somewhere into `src/backend/common`